### PR TITLE
Support color values in HSB/HSV/HSL notation

### PR DESCRIPTION
**Pull request by andreasbrett**
openHAB is using HSB/HSV as default way of sending color information. Up until now I had a conversion rule in openHAB to mitigate that but I think it's cleaner to have this conversion done by Z2M itself (especially since it is already able to take a lot of different color notations like hex, rgb etc).

This change will allow to send either HSB, HSV or HSL data:
`{ "color": { "hsl": "120,80,40" }} with hue=180°, saturation=80%, lightness=40%`
`{ "color": { "hsb": "120,80,40" }} with hue=180°, saturation=80%, brightness=40%`
`{ "color": { "hsv": "120,80,40" }} with hue=180°, saturation=80%, brightness=40%`

Separated values are also possible:
`{ "color": { "h": 120, "s": 80, "l": 40}}`
`{ "color": { "h": 120, "s": 80, "b": 40}}`
`{ "color": { "h": 120, "s": 80, "v": 40}}`

### CODE REVIEW

**Comment by sjorge** [comment]
To be [consistent](https://github.com/Koenkk/zigbee-herdsman-converters/pull/867/files#diff-3e0c275bd41af699153bf59d8dda4a2aR380) with rgb the following should also be added:
- `{ "color": { "h": 120, "s": 80, "l": 40}}`
- `{ "color": { "h": 120, "s": 80, "b": 40}}`
- `{ "color": { "h": 120, "s": 80, "v": 40}}`

**Comment by andreasbrett** [comment]
> To be [consistent](https://github.com/Koenkk/zigbee-herdsman-converters/pull/867/files#diff-3e0c275bd41af699153bf59d8dda4a2aR380) with rgb the following should also be added:
> 
> * `{ "color": { "h": 120, "s": 80, "l": 40}}`
> * `{ "color": { "h": 120, "s": 80, "b": 40}}`
> * `{ "color": { "h": 120, "s": 80, "v": 40}}`

Thing is: if I wanted to send H, S and B/V separately I can already do that like so:
`{ "brightness": 40, "color": { "hue": 120, "saturation": 80}`

> Edit: for the hue and saturation values, we should probably set them directly instead of converting to xy ? As there is a moveHueSaturation and enhanceMoveHueSaturation command in zigbee.

What exactly is the benefit of using those commands? If they're superior then we should also convert RGB, HEX etc to HSL/HSB/HSV and use them.

**Comment by sjorge** [comment]
> Thing is: if I wanted to send H, S and B/V separately I can already do that like so:
> `{ "brightness": 40, "color": { "hue": 120, "saturation": 80}`

I'm leaning more to POLA (Principle of Least Astonishment) here ,if `{"color": {"rgb": "255,255,255"}}` and `{"color": {"r": 255, "g": 255, "b": 255}}` both work, one can reasonably assume the same would hold for `{ "color": { "hsl": "120,80,40" }}` and `{ "color": { "h": 120, "s": 80, "l": 40}}`

> What exactly is the benefit of using those commands? If they're superior then we should also convert RGB, HEX etc to HSL/HSB/HSV and use them.

More of a less is more, if we can use the zigbee build in features, it's less code we have to maintain so it's a win. Although I am not sure if we can use moveHue, moveSaturation, moveHueSaturation in this case, but from glacing at the ZCL, I think it should work fine.

**Comment by andreasbrett** [comment]
> I'm leaning more to POLA (Principle of Least Astonishment) here ,if `{"color": {"rgb": "255,255,255"}}` and `{"color": {"r": 255, "g": 255, "b": 255}}` both work, one can reasonably assume the same would hold for `{ "color": { "hsl": "120,80,40" }}` and `{ "color": { "h": 120, "s": 80, "l": 40}}`

I added that to the code. Not sure it will be used but I'm with you regarding consistency.

> More of a less is more, if we can use the zigbee build in features, it's less code we have to maintain so it's a win. Although I am not sure if we can use moveHue, moveSaturation, moveHueSaturation in this case, but from glacing at the ZCL, I think it should work fine.

I'm still not happy with that as HSB->XY conversion is straight forward. If you know how to implement the moveHueSaturation plus somehow send the component (lightness or brightness depending on what has been given), please let me know.

**Comment by sjorge** [comment]
> I added that to the code. Not sure it will be used but I'm with you regarding consistency.

👍 

> I'm still not happy with that as HSB->XY conversion is straight forward. If you know how to implement the moveHueSaturation plus somehow send the component (lightness or brightness depending on what has been given), please let me know.

I don't feel super strongly about this, but to do it with moveXXXYYY commands, a bit below in the same file is an example:
https://github.com/Koenkk/zigbee-herdsman-converters/pull/867/files/27f5a30ba9f951b9b9ec0201a3599efacbd7bbd3..e8904bc6163fb683017e6ac8c50284600a117b08#diff-3e0c275bd41af699153bf59d8dda4a2aR410

It would be setting value.hue and value.saturation and the cmd to the variable.
That still leaves a separate call for brightness, which can be done with a call to ```await entity.command('genLevelCtrl', 'moveToLevel', payload, getOptions(meta));```

But lets leave that for know and see what @Koenkk things.

**Comment by Koenkk** [comment]
If the conversion is straightforward I have no problems with it.

Other thing, when openhab send the values as `hsb`/`hsv`/`hsl`, does it expect to report it back as `xy`? If not, we should do the same as https://github.com/Koenkk/zigbee-herdsman-converters/pull/867/files#diff-3e0c275bd41af699153bf59d8dda4a2aR413

**Comment by andreasbrett** [comment]
@koenkk:
But again: what do we do with the lightness / brightness value? I'm not very deep into your herdsman code and was glad I could easily put in the conversion. With regards to zcl and further zigbee stuff I'm actually quite clueless. And yes, Openhab is expecting the same format as it sends. So reporting back should be "hsb" not separate h and s values. How would we do that?

**Comment by sjorge** [comment]
@Koenkk maybe it's time for some config far (per device or global) that says what format we want the color to be returned in, 'x + y', 'r + g +b', 'rgb', ... ?

As different software might want different results.

**Comment by andreasbrett** [comment]
@sjorge
But then again why not return what has been provided. That would be POLA as well. 

@koenkk
I seem to get it now. Introduce 2 new commands `enhancedMoveToHueAndSaturationAndBrightness` and `enhancedMoveToHueAndSaturationAndLightness`.  I will try to find an easy way to convert brightness to lightness. Can't be too hard.

**Comment by sjorge** [comment]
> But then again why not return what has been provided. That would be POLA as well.

A bulb can also report a change to the color via attReport... there is no 'we set it using hsv, rgb,...' in those cases.

I assume we'd still want the published result to reflect what the software expects in that case too.

Else you will get something like

set hsv 120,80,40 on bulb_a
- triggers moveHueSaturation + moveLevel (I forgot the name for the brighhtness) being written to bulb_a
- read back result from bulb_a and respond b publishing mqtt message for bulb_a with color.hsv set.

bulb_a decided to report the change via attReport 3 seconds alter
- parse color data from bulb_a
- publish mqtt message for bulb_a with color.xy and/or color.color_temp
(I assume you want color.hsv (or whatever way you set it) here too)

**Comment by andreasbrett** [comment]
Alright, I have changed it so that hue and brightness are set via `enhancedMoveToHueAndSaturation` and brightness is set via `moveToLevelWithOnOff`. HSL is converted to HSB. No RGB or XY conversion is done anymore. Colors still look great.

File: converters/toZigbee.js [hunk 1]
```diff
@@ -391,6 +391,47 @@ const converters = {
...
+                const hsb = utils.hslToHsb(value.h, value.s, value.l);
+                newState.color = {h: hsb.h, s: hsb.s, b: hsb.b};
+                value.hue = hsb.h % 360 * (65535 / 360);
+                value.saturation = hsb.s * (2.54);
+                value.brightness = hsb.b * (2.54);
+                cmd = 'enhancedMoveToHueAndSaturationAndBrightness';
+            } else if (value.hasOwnProperty('hsl')) {
+                const hsl = value.hsl.split(',').map((i) => parseInt(i));
+                const hsb = utils.hslToHsb(hsl.h, hsl.s, hsl.l);
+                newState.color = {h: hsb.h, s: hsb.s, b: hsb.b};
```
Comment (sjorge) [inline converters/toZigbee.js]: Should this not return h, s, and l instead of h, s, and b for newState.color ?

Comment (andreasbrett) [reply converters/toZigbee.js]: good catch, fixed it

**Comment by Koenkk** [comment]
@sjorge your points from https://github.com/Koenkk/zigbee-herdsman-converters/pull/867#issuecomment-570715326 are completely valid, the ZCL is indeed vague, this probably also means that the behaviour per manufacturer is different (as they can interpet the ZCL differently). I think for now we shouldn't overcomplicate things, the controlling side is fine, reporting side indeed requires work.

**Comment by Koenkk** [comment]
@andreasbrett could you also make a PR to add some examples here? https://www.zigbee2mqtt.io/information/mqtt_topics_and_message_structure.html#zigbee2mqttfriendly_nameset

**Comment by sjorge** [comment]
> the ZCL is indeed vague
The more I read the ZCL to try and fix stuff or make things work, the more this is my general feeling. Mad props in getting zigbee-herdsman and friends to where it is now!

**Comment by andreasbrett** [comment]
> @andreasbrett could you also make a PR to add some examples here? https://www.zigbee2mqtt.io/information/mqtt_topics_and_message_structure.html#zigbee2mqttfriendly_nameset

Thanks for mergin this. PR for documentation is ready.

**Comment by Koenkk** [comment]
awesome, thanks!

**Pull request merged**